### PR TITLE
solutions/details api implemented

### PR DESCRIPTION
**Pull request by Prajwal17Tunerlabs**
# Description
These recommendations are intended to promote code quality and team communication during software development. They cover a variety of topics, including ensuring that pull requests are submitted to the correct branch, documenting new methods, preserving consistency across many services, and avoiding typical blunders like accessing APIs or DB queries within loops. Sensitive data should not be uploaded, and changes to environment variables or database models should be executed consistently. Teams may work more effectively and develop higher-quality software by adhering to these standards.

## Type of change
Please choose appropriate options.

- [ ]  Bug fix (non-breaking change which fixes an issue)
- [ ]  New feature (non-breaking change which adds functionality)
- [ ]  Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ]  Enhancement (additive changes to improve performance)
- [ ]  This change requires a documentation update

## Checklist 

- [ ]  It's critical to avoid making needless file modifications in contributions, such as adding new lines, console logs, or additional spaces, to guarantee cleaner and more efficient code. Furthermore, eliminating unnecessary imports from a file might enhance code readability and efficiency.
- [ ]  Ensure that the pull request is assigned to the right base branch and that the development branch name contains the JIRA Task Id. Furthermore, each commit message should include the JIRA Task Id in the manner "ED-100: message".
- [ ]  Only update packages if it is mentioned and authorized in the design document, and make sure that you have the required permissions.
- [ ]  Avoid making API and database queries inside a loop as it can lead to performance issues and slow down the system.
- [ ]  When calling another function inside a given function, add comments explaining the purpose and meaning of the passed arguments and expected return values.
- [ ]  If adding a blank argument in a function, add a comment explaining the reason for the blank argument.
- [ ]  Before submitting a pull request, do a self-review of your code to ensure there are no conflicts with the base branch and all comments have been addressed.
- [ ]  Before merging a pull request, it's important to have other team members review it to catch any potential errors or issues
- [ ]  To maintain code integrity, it's important to remove all related changes when removing code during a code review.
- [ ]  If new constants, endpoints, or utility functions are introduced, it is important to check if they already exist in the service to avoid any duplication.
- [ ]  Whenever a new environment variable is added to a service, it's important to ensure that the necessary changes are made to related files such as ".env.sample" and "envVariables.js" to maintain consistency and avoid errors. Additionally, the new environment variable should be added to the devops repository to ensure that it is properly documented and accessible to the team.
- [ ]  When adding a new function to a service, it is important to document it with relevant information such as the name, parameters, and return value in a consistent format across all services. Additionally, if there are any changes to the API response, ensure that the documentation in the controllers is updated accordingly. 
- [ ]  Write a clear and concise commit message that describes the changes made.
- [ ]  Maintain consistent function signature and code across all services when adding a function to multiple services. Implement changes to database models in all services that use the same model. 
- [ ]  Use only let and const. Do not use var.
- [ ]  Make common functions for repetitive code blocks.
- [ ]  Avoid uploading sensitive information such as secret tokens or passwords in pull requests to ensure data security.
- [ ]  Maintain consistent indentation and spacing throughout the code.

### CODE REVIEW

File: module/programUsers/helper.js [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+ */
+
+const programUsersQueries = require(DB_QUERY_BASE_PATH + '/programUsers')
+
+module.exports = class ProgramUsersHelper {
+	/**
+	 * check if user joined a program or not and consentShared
+	 * @method
+	 * @name checkForUserJoinedProgramAndConsentShared
+	 * @param {String} programId
```
Comment (VISHNUDAS-tunerlabs) [inline module/programUsers/helper.js:22]: @Prajwal17Tunerlabs give explanation to each param

File: module/solutions/helper.js [hunk 2]
```diff
@@ -2992,4 +2995,116 @@ module.exports = class SolutionsHelper {
 			}
 		})
 	}
+
+	/**
+	 * Fetch template observation based on solution Id.
+	 * @method
+	 * @name details
+	 * @param {String} solutionId - Solution Id.
```
Comment (VISHNUDAS-tunerlabs) [inline module/solutions/helper.js:3003]: @Prajwal17Tunerlabs Params explanations missing

File: module/solutions/helper.js [hunk 3]
```diff
@@ -2992,4 +2995,116 @@ module.exports = class SolutionsHelper {
...
+	 * @returns {Object} - Details of the solution.
+	 */
+
+	static details(solutionId, bodyData = {}, userId = '', userToken = '') {
```
Comment (VISHNUDAS-tunerlabs) [inline module/solutions/helper.js:3007]: @Prajwal17Tunerlabs why do you need userToken here?

File: module/solutions/helper.js [hunk 4]
```diff
@@ -2992,4 +2995,116 @@ module.exports = class SolutionsHelper {
...
+						}
+					}
+					templateOrQuestionDetails = await projectTemplatesHelper.details(
+						solutionData.projectTemplateId,
+						'',
+						userId,
+						isSolutionTargeted.result.isATargetedSolution ? false : true
+					)
+
+					// await improvementProjectService.getTemplateDetail(
```
Comment (VISHNUDAS-tunerlabs) [inline module/solutions/helper.js:3041]: @Prajwal17Tunerlabs remove unused code

**Review by VISHNUDAS-tunerlabs — CHANGES_REQUESTED** [review]
reviewed on 11 june 2024, 9:30

File: module/solutions/helper.js (hunk 4 above)
Comment (Prajwal17Tunerlabs) [reply module/solutions/helper.js:3041]: removed the unused code

File: module/solutions/helper.js (hunk 3 above)
Comment (Prajwal17Tunerlabs) [reply module/solutions/helper.js:3007]: unused token removed form here

File: module/solutions/helper.js (hunk 2 above)
Comment (Prajwal17Tunerlabs) [reply module/solutions/helper.js:3003]: params explanation added

File: module/programUsers/helper.js (hunk 1 above)
Comment (Prajwal17Tunerlabs) [reply module/programUsers/helper.js:22]: params explanation added

File: controllers/v1/solutions.js [hunk 5]
```diff
@@ -1011,4 +1011,124 @@ module.exports = class Solutions extends Abstract {
...
+	 * @name details
+	 * @param {Object} req - requested data.
+	 * @param {String} req.params._id - solution Id
+	 * @returns {Array}
+	 */
+
+	async details(req) {
+		return new Promise(async (resolve, reject) => {
+			try {
+				let solutionData = await solutionsHelper.details(
```
[the remark this answers is not in the document]
Comment (Prajwal17Tunerlabs) [reply controllers/v1/solutions.js:1118]: params explanation added

File: models/solutions.js [hunk 6]
```diff
@@ -5,90 +5,94 @@
...
-        default: 1
-      },
-      reportInformation : Object,
-      certificateTemplateId : "ObjectId",
-      rootOrganisations : Array,
-      createdFor : Array
-    }
-  };
\ No newline at end of file
+	name: 'solutions',
```
[the remark this answers is not in the document]
Comment (Prajwal17Tunerlabs) [reply models/solutions.js:8]: yes

`projectTemplateId: {
	type: 'ObjectId',
	index: true,
},`

was added

File: module/project/templates/helper.js [hunk 7]
```diff
@@ -924,7 +926,8 @@ module.exports = class ProjectTemplatesHelper {
 				// }
 
 				if (templateId) {
-					let validateTemplateId = UTILS.isValidMongoId(templateId)
+					console.log(templateId)
```
[the remark this answers is not in the document]
Comment (Prajwal17Tunerlabs) [reply module/project/templates/helper.js:929]: console removed

File: controllers/v1/solutions.js [hunk 8]
```diff
@@ -1011,4 +1011,124 @@ module.exports = class Solutions extends Abstract {
...
+	/**
+	 * get solution details
+	 * @method
+	 * @name details
+	 * @param {Object} req - requested data.
+	 * @param {String} req.params._id - solution Id
+	 * @returns {Array}
+	 */
+
+	async details(req) {
```
[the remark this answers is not in the document]
Comment (Prajwal17Tunerlabs) [reply controllers/v1/solutions.js:1115]: validations are now added to solution/details api

**Pull request merged**